src/vs/workbench/contrib/customEditor/browser/extensionContributedEditorService.ts

```diff
@@ -0,0 +1,343 @@
+		// const strictMatchEditorInput = group.editors.find(e => e === editor && !this._fileEditorInputFactory.isFileEditorInput(e));
+		// if (strictMatchEditorInput) {
+		// 	return;
+		// }
+
+		// If an existing editor for a resource exists within the group and we're reopening it, replace it.
+		const existing = firstOrDefault(this.editorService.findEditors(resource, group));
+		if (existing) {
+			if (!input.matches(existing)) {
+				await this.editorService.replaceEditors([{
+					editor: existing,
+					replacement: input,
+					forceReplaceDirty: existing.resource?.scheme === Schemas.untitled,
+					options: options ? EditorOptions.create(options) : undefined,
+				}], group);
+
+				if (selectedContribution.editorInfo.instanceOf(existing)) {
+					existing.dispose();
+				}
+			}
+		}
+		return group.openEditor(input, options);
+	}
+
+	private closeExistingEditorsForResource(
+		resource: URI,
+		viewType: string,
+		targetGroup: IEditorGroup,
+	): void {
+		const editorInfoForResource = this.findExistingEditorsForResource(resource, viewType);
+		if (!editorInfoForResource.length) {
+			return;
+		}
+
+		const editorToUse = editorInfoForResource[0];
+
+		// Replace all other editors
+		for (const { editor, group } of editorInfoForResource) {
+			if (editor !== editorToUse.editor) {
+				group.closeEditor(editor);
+			}
+		}
+
+		if (targetGroup.id !== editorToUse.group.id) {
+			editorToUse.group.closeEditor(editorToUse.editor);
+		}
+		return;
+	}
+
+	/**
+	 * Given a resource and a viewType, returns all editors open for that resouce and viewType.
+	 * @param resource The resource specified
+	 * @param viewType The viewtype
+	 * @returns A list of editors
+	 */
+	private findExistingEditorsForResource(
+		resource: URI,
+		viewType: string,
+	): Array<{ editor: IEditorInput, group: IEditorGroup }> {
+		const out: Array<{ editor: IEditorInput, group: IEditorGroup }> = [];
+		const orderedGroups = distinct([
+			...this.editorGroupService.groups,
+		]);
+
+		for (const group of orderedGroups) {
+			for (const editor of group.editors) {
+				if (isEqual(editor.resource, resource) && (editor as IContributedEditorInput).viewType === viewType) {
+					out.push({ editor, group });
+				}
+			}
+		}
+		return out;
+	}
+
+	// @ts-ignore
+	private storeUserChoice(override: string) {
+		const currentChoices: ExtensionContributedEditorChoiceEntry[] = JSON.parse(this.storageService.get(this._editorChoiceStorageID, StorageScope.GLOBAL, '[]'));
+		const currentChoice = currentChoices.find(entry => entry.editorID === override);
+		if (!currentChoice) {
+			currentChoices.push({ editorID: override, choice: ExtensionContributedEditorChoice.OPTIONAL });
+			this.storageService.store(this._editorChoiceStorageID, JSON.stringify(currentChoices), StorageScope.GLOBAL, StorageTarget.USER);
+		}
+	}
+}
+
+export const IExtensionContributedEditorService = createDecorator<IExtensionContributedEditorService>('extensionContributedEditorService');
```

Comment:
I find the name `IExtensionContributedEditorService` somewhat misleading because the service is not an `IEditorService` in the end. It is a service to register editor overrides with. Should the name reflect that better? Instead of "extension contributed editors service", maybe more along the lines of the intent:
* you can register editors to be used based on certain rules
* the editors are used whenever the editor service opens an editor
* this is agnostic to extensions right, so why have "extensions" in the name?

Maybe more like "editor override service"?

Reply:
I'm hesitant to call it editor override service in case there is more shared code that could live here, i.e. what if we wanted to support what an editor does when being moved. We could also say this service is just for opening and nothing else to prevent it from growing out of scope and then just called it EditorOverrideService

Reply:
How about "custom editor provider service" given it seems to reside in the `contrib/custom` world for now?

Reply:
Should we keep it residing in the contrib/custom world? I just put it there, I thought it might move to be near the editor service but wanted your input on that

Reply:
I think it should move closer to the editor service, but that means to move anything associated with it too because you cannot depend from that core layer to a contributed layer (like `custom` is).

For the purpose of this PR, I don't mind for now not moving the files and look at that only later.

Reply:
After discussion we will rename it to EditorOverrideService and move it in a separate PR.